x/consensus/keeper/estimate.go

```diff
@@ -0,0 +1,162 @@
+				if err := k.checkAndProcessEstimatedMessage(ctx, msg, cq); err != nil {
+					logger.WithError(err).Warn("Failed to process estimated message")
+					continue
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (k Keeper) checkAndProcessEstimatedMessage(ctx context.Context,
+	msg types.QueuedSignedMessageI,
+	q consensus.Queuer,
+) error {
+	_, rcid := q.ChainInfo()
+	logger := liblog.FromKeeper(ctx, k).WithComponent("check-and-process-estimated-messages").WithChain(rcid)
+
+	// Skip messages that don't require gas estimation
+	if !msg.GetRequireGasEstimation() {
+		return nil
+	}
+
+	// Skip messages that don't have gas estimates
+	if len(msg.GetGasEstimates()) < 1 {
+		return nil
+	}
+	logger = logger.WithFields(
+		"id", msg.GetId(),
+		"nonce", msg.Nonce())
+	logger.Debug("Processing gas estimates for message.")
+
+	estimate, err := k.consensusChecker.VerifyGasEstimates(ctx, k,
+		slice.Map(msg.GetGasEstimates(), func(ge *types.GasEstimate) libcons.GasEstimate {
+			return ge
+		}))
+	if err != nil {
+		if errors.Is(err, libcons.ErrConsensusNotAchieved) {
+			logger.WithError(err).Debug("Consensus not achieved for gas estimates")
+			return nil
+		}
+		return fmt.Errorf("failed to verify gas estimates: %w", err)
+	}
+
+	if err := q.SetElectedGasEstimate(ctx, msg.GetId(), estimate); err != nil {
+		return fmt.Errorf("failed to set elected gas estimate: %w", err)
+	}
+
+	if err := k.checkAndProcessEstimatedSubmitLogicCall(ctx, msg, q, estimate); err != nil {
+		return fmt.Errorf("failed to process estimated submit logic call: %w", err)
+	}
+
+	return nil
+}
+
+func (k Keeper) checkAndProcessEstimatedSubmitLogicCall(
+	ctx context.Context,
+	msg types.QueuedSignedMessageI,
+	q consensus.Queuer,
+	estimate uint64,
+) error {
+	m, err := libmsg.ToEvmMessage(msg, k.cdc)
+	if err != nil {
+		return fmt.Errorf("failed to convert message to evm message: %w", err)
+	}
+	action, ok := m.Action.(*evmtypes.Message_SubmitLogicCall)
+	if !ok {
+		// Skip messages that are not SubmitLogicCall
+		return nil
+	}
+
+	fees, err := k.calculateFeesForEstimate(ctx, sdk.ValAddress(m.GetAssignee()), m.GetChainReferenceID(), estimate)
+	if err != nil {
+		return fmt.Errorf("failed to calculate fees for estimate: %w", err)
+	}
+	action.SubmitLogicCall.Fees = fees
+	_, err = q.Put(ctx, msg, &consensus.PutOptions{
+		MsgIDToReplace: msg.GetId(),
+	})
+
+	return err
+}
+
+func (k Keeper) calculateFeesForEstimate(
+	ctx context.Context,
+	relayer sdk.ValAddress,
+	chainReferenceID string,
+	estimate uint64,
+) (*evmtypes.SubmitLogicCall_Fees, error) {
+	fees := &evmtypes.SubmitLogicCall_Fees{}
+	multiplicators, err := k.feeProvider(ctx, relayer, chainReferenceID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get fee settings: %w", err)
+	}
+
+	// Relayer fees are stored as multiplicator of total gas estimate value
+	fees.RelayerFee = multiplicators.RelayerFee.
+		MulInt(math.NewIntFromUint64(estimate)).
+		Ceil().
+		BigInt().
+		Uint64()
```

Comment:
Do we need all these int type conversions, or could we do it directly in uint64? The final result has to fit in uint64, so what are we gaining by multiplying with big types?

Reply:
Since `float64` is not deterministic, the `multiplicators` are stored as `LegacyDec`.  The casts are needed to work against the type interface unfortunately.